libfastpm/thermalvelocity.c

```diff
@@ -73,29 +73,52 @@ void pix2vec (int pix, double *vec, int n_side)
     vec[2] = z;
 }
 
+/* make params struct for functions we will integrate with gsl */
+typedef struct kernel_params {double m[3]; int n;} kernel_params;
 
-// Functional form of F-D distribution
-double fermi_dirac_kernel(double x, void * params)
+double fermi_dirac_kernel_vol_1(double x)
 {
-    return x * x / (exp(x) + 1);
+    /* Functional form of FD distribution for 1 particle
+       over the full volume (i.e. no *x^2) */
+    return 1. / (exp(x) + 1);
 }
 
-// Samples the low velocity end more effectively
-double low_vel_kernel(double x, void * params)
+double fermi_dirac_kernel_vol(double x, void * p)
 {
-    return x / (exp(x) + 1);
+    /* FD kernel in the case of multiple particles
+       over the full volume (i.e. no *x^2) */
+    struct kernel_params * params = (struct kernel_params *)p;
+
+    double result = fermi_dirac_kernel_vol_1(x);
+    // only loop from i=1. i=0 term has been included in result already (r=1)
+    for(int i = 1; i < params->n; i ++) {
+        double r = params->m[i] / params->m[0];
+        if (r > 1) fastpm_raise(-1, "Please input the heaviest ncdm particle first.\n");
+        result += (r*r*r*r) * fermi_dirac_kernel_vol_1(x * r);
+    }
+    return result;
 }
 
-// Needed for calculatin the velocity dispersion
-double fermi_dirac_dispersion(double x, void * params)
+/* now define the 1D kernel and moments */
+double fermi_dirac_kernel(double x, void * p)
 {
-    return x * x * x * x/ (exp(x) + 1);
+    return x*x * fermi_dirac_kernel_vol(x, p);
 }
 
+/* Samples the low velocity end more effectively */
+double low_vel_kernel(double x, void * p)
+{
+    return x * fermi_dirac_kernel_vol(x, p);
+}
 
+/* Needed for calculatin the velocity dispersion */
+double fermi_dirac_dispersion(double x, void * p)
+{
+     return x*x*x*x * fermi_dirac_kernel_vol(x, p);
+}
 
 
-void divide_fd(double *vel_table, double *mass, int n_shells, int lvk)
+void divide_fd(double *vel_table, double *mass, kernel_params * params, int n_shells, int lvk)
```

Comment:
Long functions you can reformat it by keeping 'void' as a line, break lines of function arguments at meaning group boundaries.